README.md
# werewolf-slack

[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![Build Status](https://travis-ci.org/hjwylde/werewolf-slack.svg?branch=master)](https://travis-ci.org/hjwylde/werewolf-slack)
[![Release](https://img.shields.io/github/release/hjwylde/werewolf-slack.svg)](https://github.com/hjwylde/werewolf-slack/releases/latest)
[![werewolf-slack on Stackage LTS](https://www.stackage.org/package/werewolf-slack/badge/lts)](https://www.stackage.org/lts/package/werewolf-slack)
[![werewolf-slack on Stackage Nightly](https://www.stackage.org/package/werewolf-slack/badge/nightly)](https://www.stackage.org/nightly/package/werewolf-slack)

A chat interface for playing [werewolf](https://github.com/hjwylde/werewolf) in
    [Slack](https://slack.com/).
Werewolf is a well known social party game, commonly also called Mafia.
See the [Wikipedia article](https://en.wikipedia.org/wiki/Mafia_(party_game)) for a rundown on its
    gameplay and history.

![Werewolf for Slack](https://raw.githubusercontent.com/hjwylde/werewolf-slack/master/.readme/werewolf-for-slack.png)

### Game description

Long has the woods been home to wild creatures, both kind and cruel.
Most have faces and are known to the inhabitants of Fougères in Brittany, France; but no-one from
    the village has yet to lay eyes on the merciless Werewolf.

Each night Werewolves attack the village and devour the innocent.
For centuries no-one knew how to fight this scourge, however recently a theory has taken ahold that
    mayhaps the Werewolves walk among the Villagers themselves...

Objective of the game:  
For the Loners: complete their own objective.  
For the Villagers: lynch all of the Werewolves.  
For the Werewolves: devour all of the Villagers.

### Setup

#### Preparing Slack

Set up an Incoming Webhook [here](https://my.slack.com/services/new/incoming-webhook/).
Make note of the *webhook url*, we'll be using that soon.

Set up a Slash Command (`/werewolf` or similar)
    [here](https://my.slack.com/services/new/slash-commands/).
The Slash Command should perform a GET request to the server werewolf-slack is going to be hosted
    on.
Make note of the *token* here too.

#### Installing

Installing werewolf-slack is easiest done using either
    [Docker](https://www.docker.com/) (recommended),
    [stack](https://github.com/commercialhaskell/stack) or
    [Cabal](https://github.com/haskell/cabal).

**Using Docker:**

```bash
docker pull hjwylde/werewolf-slack-docker
```

**Using stack:**

```bash
stack install werewolf-slack werewolf
export PATH=$PATH:~/.local/bin
```

**Using Cabal:**

```bash
cabal-install werewolf-slack werewolf
export PATH=$PATH:~/.cabal/bin
```

#### Running

werewolf-slack is a simple web server that listens for events from the Slack Slash Command.
After receiving an event werewolf-slack forwards it on to the werewolf game engine and uses the
    Incoming Webhook to send back the response.

Running werewolf-slack is easiest done using either
    [Docker](https://www.docker.com/) (recommended) or
    the binary itself.
Make sure to add rules to your firewall for werewolf-slack's port.

**With Docker:**

```bash
docker run -d -p 80:80 hjwylde/werewolf-slack-docker -v TOKEN -u WEBHOOK_URL
```

**With werewolf-slack:**

```bash
werewolf-slack -v TOKEN -u WEBHOOK_URL &
```

#### Configuration

It is possible to also configure the port that werewolf-slack listens on. This is done via the
    `--port` (`-p`) option.

The default port is *80*.

### Usage

Type `/werewolf help` in your Slack channel to get going!


CHANGELOG.md
# Changelog

### Upcoming

### v1.0.2.0

*Revisions*

* Bumped max version constraint of optparse-applicative. ([#20](https://github.com/hjwylde/werewolf/issues/20))

### v1.0.1.5

*Revisions*

* Bumped max version constraint of werewolf.

### v1.0.1.4

*Revisions*

* Bumped max version constraint of werewolf.

### v1.0.1.3

*Revisions*

* Bumped max version constraint of werewolf.

### v1.0.1.2

*Revisions*

* Bumped max version constraint of werewolf.

### v1.0.1.1

*Revisions*

* Bumped max version constraint of werewolf. ([#18](https://github.com/hjwylde/werewolf/issues/18))

### v1.0.1.0

*Revisions*

* Changed the default port to 80.
* Removed the binary header and updated the program description.

### v1.0.0.0

*Major*

* Initial stable release!

### v0.4.0.0

*Major*

* Renamed `--authentication-token` to `--webhook-url`. ([#15](https://github.com/hjwylde/werewolf/issues/15))
* Renamed `--validation-token` to `--token`. ([#15](https://github.com/hjwylde/werewolf/issues/15))
* Removed `--channel-name` option. ([#16](https://github.com/hjwylde/werewolf/issues/16))

*Minor*

* Split out Dockerfile into its own repository. ([#13](https://github.com/hjwylde/werewolf/issues/13))

*Revisions*

* Changed 202 response to 200. ([#17](https://github.com/hjwylde/werewolf/issues/17))

### v0.3.0.2

*Revisions*

* Removed werewolf version specification from stack.yaml.

### v0.3.0.1

*Revisions*

* Added version to werewolf in Dockerfile.

### v0.3.0.0

*Major*

* Added support for werewolf v0.5.0.0. ([#12](https://github.com/hjwylde/werewolf/issues/12))

### v0.2.0.2

*Revisions*

* Added version to werewolf in Dockerfile.

### v0.2.0.1

*Revisions*

* Added version to werewolf-slack in Dockerfile. ([#11](https://github.com/hjwylde/werewolf/issues/11))

### v0.2.0.0

*Major*

* Added options for specifying the Slack access token, channel name and port to run the server on. ([#6](https://github.com/hjwylde/werewolf/issues/6))

*Minor*

* Added a `--validation-token` option and validation. ([#9](https://github.com/hjwylde/werewolf/issues/9))
* Added a `--debug` flag. ([#4](https://github.com/hjwylde/werewolf/issues/4))

*Revisions*

* Fingers crossed fixed a bug with UTF-8 parsing of an accented e. ([#10](https://github.com/hjwylde/werewolf/issues/10))

### v0.1.0.0

*Major*

* Initial version that connects Slack with the werewolf binary. ([#1](https://github.com/hjwylde/werewolf/issues/1))

*Minor*

* Added a Dockerfile. ([#5](https://github.com/hjwylde/werewolf/issues/5))


app/Werewolf/Slack/Slack.hs
{-|
Module      : Werewolf.Slack.Slack

Copyright   : (c) Henry J. Wylde, 2016
License     : BSD3
Maintainer  : public@hjwylde.com
-}

{-# LANGUAGE FlexibleContexts      #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings     #-}

module Werewolf.Slack.Slack (
    -- * Slack
    notify,
) where

import Control.Monad.Extra
import Control.Monad.Reader
import Control.Monad.State

import Data.Aeson

import Network.HTTP.Client
import Network.HTTP.Types.Method

import Werewolf.Slack.Options

notify :: (MonadIO m, MonadReader Options m, MonadState Manager m) => Maybe String -> String -> m ()
notify mTo message = do
    manager <- get

    initialRequest  <- asks optWebhookUrl >>= liftIO . parseUrl
    let request     = initialRequest { method = methodPost, requestBody = body }

    whenM (asks optDebug) $ liftIO (print request)

    response <- liftIO $ httpLbs request manager

    whenM (asks optDebug) $ liftIO (print response)
    where
        body    = RequestBodyLBS $ encode payload
        payload = object
            [ "channel" .= mTo
            , "text" .= message
            ]


app/Werewolf/Slack/Werewolf.hs
{-|
Module      : Werewolf.Slack.Werewolf

Copyright   : (c) Henry J. Wylde, 2016
License     : BSD3
Maintainer  : public@hjwylde.com
-}

{-# LANGUAGE FlexibleContexts      #-}
{-# LANGUAGE MultiParamTypeClasses #-}

module Werewolf.Slack.Werewolf (
    -- * Werewolf
    execute,
) where

import Control.Monad.Extra
import Control.Monad.Reader
import Control.Monad.State

import           Data.Aeson
import qualified Data.Text               as T
import qualified Data.Text.Lazy          as TL
import           Data.Text.Lazy.Encoding

import Game.Werewolf

import Network.HTTP.Client hiding (Response)

import System.Process

import Werewolf.Slack.Options
import Werewolf.Slack.Slack

execute :: (MonadIO m, MonadReader Options m, MonadState Manager m) => String -> String -> String -> m ()
execute tag user userCommand = whenJustM (interpret tag user userCommand) handle

interpret :: MonadIO m => String -> String -> String -> m (Maybe Response)
interpret tag user userCommand = do
    stdout <- liftIO $ readCreateProcess (proc command arguments) ""

    return (decode (encodeUtf8 $ TL.pack stdout) :: Maybe Response)
    where
        atUser      = if take 1 user == "@" then user else '@':user
        command     = "werewolf"
        arguments   = ["--caller", atUser, "--tag", tag, "interpret", "--"] ++ words userCommand

handle :: (MonadIO m, MonadReader Options m, MonadState Manager m) => Response -> m ()
handle response = do
    whenM (asks optDebug) $ liftIO (print response)

    forM_ (messages response) $ \(Message mTo message) ->
        notify (T.unpack <$> mTo) (T.unpack message)


app/Werewolf/Slack/Application.hs
{-|
Module      : Werewolf.Slack.Application

Copyright   : (c) Henry J. Wylde, 2016
License     : BSD3
Maintainer  : public@hjwylde.com
-}

{-# LANGUAGE FlexibleContexts      #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings     #-}

module Werewolf.Slack.Application (
    -- * Application
    runApplication,
) where

import Control.Concurrent
import Control.Monad.Extra
import Control.Monad.Reader
import Control.Monad.State

import qualified Data.ByteString.Char8      as BSC
import qualified Data.ByteString.Lazy.Char8 as BSLC
import           Data.Maybe

import Network.HTTP.Client      hiding (queryString)
import Network.HTTP.Client.TLS
import Network.HTTP.Types
import Network.Wai              hiding (Response, requestBody, responseStatus)
import Network.Wai.Handler.Warp

import Werewolf.Slack.Options
import Werewolf.Slack.Werewolf

runApplication :: (MonadIO m, MonadReader Options m) => m ()
runApplication = do
    options <- ask

    manager <- liftIO $ newManager tlsManagerSettings

    liftIO $ run (optPort options) (application options manager)

application :: Options -> Manager -> Application
application options manager request respond
    | isNothing mToken                          = debugRequest >> badRequest
    | fromJust mToken /= optToken options       = debugRequest >> unauthorized
    | isNothing mUser || isNothing mUserCommand = debugRequest >> badRequest
    | otherwise                                 = debugRequest >> forkIO (evalStateT (runReaderT action options) manager) >> ok
    where
        debugRequest = when (optDebug options) $ print request

        ok              = respond $ responseLBS status200 [] (BSLC.pack $ unwords [":wolf:", fromJust mUserCommand, ":moon:"])
        badRequest      = respond $ responseLBS status400 [] "bad request"
        unauthorized    = respond $ responseLBS status401 [] "unauthorized"

        param name      = join . lookup name $ queryString request
        mToken          = BSC.unpack <$> param "token"
        mUser           = BSC.unpack <$> param "user_name"
        mUserCommand    = BSC.unpack <$> param "text"
        action          = fromJust $ execute <$> mToken <*> mUser <*> mUserCommand


app/Werewolf/Slack/Options.hs
{-|
Module      : Werewolf.Slack.Options
Description : Optparse utilities.

Copyright   : (c) Henry J. Wylde, 2016
License     : BSD3
Maintainer  : public@hjwylde.com

Optparse utilities.
-}

{-# LANGUAGE CPP #-}

module Werewolf.Slack.Options (
    -- * Options
    Options(..),

    -- * Optparse
    werewolfSlackPrefs, werewolfSlackInfo, werewolfSlack,
) where

#if MIN_VERSION_optparse_applicative(0,13,0)
import Data.Monoid
#endif
import Data.Version (showVersion)

import Network.Wai.Handler.Warp

import Options.Applicative

import qualified Werewolf.Slack.Version as This

data Options = Options
    { optDebug      :: Bool
    , optPort       :: Port
    , optToken      :: String
    , optWebhookUrl :: String
    } deriving (Eq, Show)

-- | The default preferences.
--   Limits the help output to 100 columns.
werewolfSlackPrefs :: ParserPrefs
#if MIN_VERSION_optparse_applicative(0,13,0)
werewolfSlackPrefs = prefs $ columns 100 <> showHelpOnEmpty
#else
werewolfSlackPrefs = prefs $ columns 100
#endif

-- | An optparse parser of a werewolf-slack command.
werewolfSlackInfo :: ParserInfo Options
werewolfSlackInfo = info (infoOptions <*> werewolfSlack) (fullDesc <> progDesc')
    where
        infoOptions = helper <*> version
        version     = infoOption ("Version " ++ showVersion This.version) $ mconcat
            [ long "version", short 'V', hidden
            , help "Show this binary's version"
            ]

        progDesc' = progDesc "Runs a simple web server that links the werewolf game engine and Slack integrations"

-- | An options parser.
werewolfSlack :: Parser Options
werewolfSlack = Options
    <$> switch (mconcat
        [ long "debug", short 'd'
        , help "Enable debug mode"
        ])
    <*> portOption (mconcat
        [ long "port", short 'p', metavar "NAT"
        , value 80, showDefault
        , help "Specify the port for the server to listen on"
        ])
    <*> strOption (mconcat
        [ long "token", short 't', metavar "TOKEN"
        , help "Specify the Slash Command token"
        ])
    <*> strOption (mconcat
        [ long "webhook-url", short 'u', metavar "URL"
        , help "Specify the Incoming Webhook URL"
        ])
    where
        portOption = option auto
